relay-spans/src/v1_to_v2.rs

```diff
@@ -0,0 +1,428 @@
+        .map(|link| {
+            link.map_value(
+                |SpanLink {
+                     trace_id,
+                     span_id,
+                     sampled,
+                     attributes,
+                     other,
+                 }| {
+                    SpanV2Link {
+                        trace_id,
+                        span_id,
+                        sampled,
+                        attributes: attributes.map_value(|attrs| {
+                            Attributes::from_iter(
+                                attrs
+                                    .into_iter()
+                                    .map(|(key, value)| (key, attribute_from_value(value))),
+                            )
+                        }),
+                        other,
+                    }
+                },
+            )
+        })
+        .collect()
+}
+
+fn attributes_from_data(data: Annotated<SpanData>) -> Annotated<Attributes> {
+    let Annotated(data, meta) = data;
+    let Some(data) = data else {
+        return Annotated(None, meta);
+    };
+    let Value::Object(data) = data.into_value() else {
+        debug_assert!(false, "`SpanData` must convert to Object");
+        return Annotated(None, meta);
+    };
+
+    Annotated::new(Attributes::from_iter(data.into_iter().filter_map(
+        |(key, value)| (!value.is_empty()).then_some((key, attribute_from_value(value))),
+    )))
+}
+
+/// Converts a generic [`Value`] into an annotated attribute with the proper type.
+///
+/// - Any conversion errors are documented in [`Meta`].
+/// - Nested values are serialized into strings.
+fn attribute_from_value(value: Annotated<Value>) -> Annotated<Attribute> {
+    let value: Annotated<AttributeValue> =
+        value.and_then(|value| attribute_value_from_value(value));
+    value.map_value(Attribute::from)
+}
+
+fn attribute_value_from_value(value: Value) -> Annotated<AttributeValue> {
+    match value {
+        Value::Bool(v) => AttributeValue::from(v),
+        Value::I64(v) => AttributeValue::from(v),
+        Value::U64(v) => match i64::try_from(v) {
+            Ok(i) => AttributeValue::from(i),
+            Err(_) => return Annotated::from_error(Error::invalid("integer too large"), None),
+        },
+        Value::F64(v) => AttributeValue::from(v),
+        Value::String(v) => AttributeValue::from(v),
+        Value::Array(_) | Value::Object(_) => {
+            return match Annotated::new(value).to_json() {
+                Ok(s) => Annotated(
+                    Some(AttributeValue {
+                        ty: AttributeType::String.into(),
+                        value: Value::String(s).into(),
+                    }),
+                    Meta::from_error(Error::expected("scalar attribute")),
+                ),
+                Err(_) => Annotated::from_error(
+                    Error::invalid("failed to serialize nested attribute"),
+                    None,
+                ),
+            };
+        }
```

Comment:
We already stringify nested `data` fields [in the segment consumer](https://github.com/getsentry/sentry/blob/ba109b3c503d98798037ff0714207f2a336aa6c2/src/sentry/spans/consumers/process_segments/convert.py#L98-L99), so this is not be a breaking for the system end-to-end.